File: services/ki-usability/app/schemas/voice.py
"""
Voice resolve schemas
"""

from typing import Any, Dict, Optional
from pydantic import BaseModel, Field


class VoiceResolveIn(BaseModel):
    """Input for voice-to-intent resolution."""

    text: str = Field(..., description="Transcribed text (e.g. from Web Speech API)")
    context: Optional[Dict[str, Any]] = Field(
        default_factory=dict,
        description="Context: domain, mask, tenant_id",
    )


class VoiceResolveOut(BaseModel):
    """Resolved intent from voice text."""

    action_id: str = Field(..., description="Resolved action ID")
    params: Dict[str, Any] = Field(default_factory=dict, description="Extracted parameters")
    confidence: float = Field(..., ge=0, le=1, description="Confidence 0..1")
    raw_text: str = Field(..., description="Original input text")


class VoicePolishIn(BaseModel):
    """Input for transcript polishing."""

    text: str = Field(..., min_length=1, description="Raw STT transcript")
    tone: str = Field(default="business", description="business | casual | formal")


class VoicePolishOut(BaseModel):
    """Polished transcript."""

    raw_text: str
    polished_text: str
    provider: str = Field(..., description="ollama:model | passthrough | fallback-raw")
    polish_applied: bool = False
    error: Optional[str] = None


class VoiceTranscribeIn(BaseModel):
    """Audio upload for local STT (base64-encoded)."""

    audio_base64: str = Field(..., description="Base64-encoded audio (wav/webm/mp3)")
    audio_format: str = Field(default="wav", description="wav | webm | mp3 | ogg")
    language: str = Field(default="de-DE")


class VoiceTranscribeOut(BaseModel):
    """STT result."""

    text: str
    confidence: float = 0.0
    provider: str = "faster-whisper"
    language: str = "de-DE"
    error: Optional[str] = None


class VoiceStatusOut(BaseModel):
    """Voice stack readiness."""

    stt_provider: str
    stt_ready: bool
    tts_provider: str
    tts_ready: bool
    ollama_base_url: str
    ollama_reachable: bool
    voice_polish_enabled: bool
    voice_summary_enabled: bool
    kokoro_configured: bool = False
    piper_configured: bool = False
    faster_whisper_model: str = "small"


class VoiceSummaryIn(BaseModel):
    """Input for voice summary."""

    text: str = Field(..., min_length=1, description="Source text to summarize")
    max_words: Optional[int] = Field(default=None, ge=10, le=120)


class VoiceSummaryOut(BaseModel):
    """Speakable summary (~15s)."""

    source_text: str
    summary_text: str
    provider: str
    summary_applied: bool = False
    estimated_seconds: int = 0
    error: Optional[str] = None


class VoiceSynthesizeIn(BaseModel):
    """Input for local TTS."""

    text: str = Field(..., min_length=1, description="Text to speak")
    language: str = Field(default="de-DE")


class VoiceSynthesizeOut(BaseModel):
    """TTS result — Piper audio or browser hint."""

    text: str
    provider: str
    browser_hint: bool = False
    audio_base64: Optional[str] = None
    content_type: Optional[str] = None
    language: str = "de-DE"
    hint: Optional[str] = None
    error: Optional[str] = None


class VoicePipelineIn(BaseModel):
    """WhisperBar unified pipeline input."""

    mode: str = Field(..., description="dictate | summary | intent")
    text: Optional[str] = Field(default=None, description="Text or clipboard content")
    audio_base64: Optional[str] = Field(default=None, description="Optional STT audio")
    audio_format: str = Field(default="wav")
    language: str = Field(default="de-DE")
    tone: str = Field(default="business")
    context: Optional[Dict[str, Any]] = Field(default_factory=dict)
    synthesize: bool = Field(default=False, description="TTS for summary mode")


class VoicePipelineOut(BaseModel):
    """WhisperBar pipeline result (mode-specific fields)."""

    mode: str
    raw_text: Optional[str] = None
    polished_text: Optional[str] = None
    source_text: Optional[str] = None
    summary_text: Optional[str] = None
    provider: Optional[str] = None
    polish_applied: Optional[bool] = None
    summary_applied: Optional[bool] = None
    estimated_seconds: Optional[int] = None
    text: Optional[str] = None
    resolved: Optional[Dict[str, Any]] = None
    tts: Optional[Dict[str, Any]] = None
    error: Optional[str] = None

File: packages/frontend-web/src/features/ki-usability/api/voice.ts
/**
 * Voice resolve: send transcribed text to ki-usability-api, get action_id + params
 */

export interface VoiceResolveRequest {
  text: string
  context?: { domain?: string; mask?: string; tenant_id?: string }
}

export interface VoiceResolveResponse {
  action_id: string
  params: Record<string, unknown>
  confidence: number
  raw_text: string
}

export interface VoicePolishRequest {
  text: string
  tone?: 'business' | 'casual' | 'formal'
}

export interface VoicePolishResponse {
  raw_text: string
  polished_text: string
  provider: string
  polish_applied: boolean
  error?: string
}

export interface VoiceSummaryRequest {
  text: string
  max_words?: number
}

export interface VoiceSummaryResponse {
  source_text: string
  summary_text: string
  provider: string
  summary_applied: boolean
  estimated_seconds: number
  error?: string
}

export interface VoiceSynthesizeRequest {
  text: string
  language?: string
}

export interface VoiceSynthesizeResponse {
  text: string
  provider: string
  browser_hint: boolean
  audio_base64?: string | null
  content_type?: string | null
  language?: string
  hint?: string
  error?: string
}

export interface VoicePipelineRequest {
  mode: 'dictate' | 'summary' | 'intent'
  text?: string
  audio_base64?: string
  audio_format?: string
  language?: string
  tone?: 'business' | 'casual' | 'formal'
  context?: { domain?: string; mask?: string; tenant_id?: string }
  synthesize?: boolean
}

export interface VoicePipelineResponse {
  mode: string
  raw_text?: string
  polished_text?: string
  source_text?: string
  summary_text?: string
  provider?: string
  polish_applied?: boolean
  summary_applied?: boolean
  estimated_seconds?: number
  text?: string
  resolved?: {
    action_id: string
    params: Record<string, unknown>
    confidence: number
    raw_text: string
  } | null
  tts?: VoiceSynthesizeResponse
  error?: string
}

export interface VoiceTranscribeRequest {
  audio_base64: string
  audio_format?: 'wav' | 'webm' | 'mp3' | 'ogg'
  language?: string
}

export interface VoiceTranscribeResponse {
  text: string
  confidence: number
  provider: string
  language: string
  error?: string
}

/** Base URL for ki-usability-api (dev: use Vite proxy /api/ki-usability → localhost:5200) */
const BASE =
  (import.meta.env as Record<string, string | undefined>).VITE_KI_USABILITY_API_URL ?? '/api/ki-usability'

export async function resolveVoice(
  body: VoiceResolveRequest
): Promise<VoiceResolveResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/resolve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) return null
  const data = await res.json()
  if (data == null) return null
  return data as VoiceResolveResponse
}

export async function polishVoice(body: VoicePolishRequest): Promise<VoicePolishResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/polish`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ text: body.text, tone: body.tone ?? 'business' }),
  })
  if (!res.ok) return null
  return (await res.json()) as VoicePolishResponse
}

export async function summarizeVoice(body: VoiceSummaryRequest): Promise<VoiceSummaryResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/summary`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) return null
  return (await res.json()) as VoiceSummaryResponse
}

export async function synthesizeVoice(body: VoiceSynthesizeRequest): Promise<VoiceSynthesizeResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/synthesize`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ text: body.text, language: body.language ?? 'de-DE' }),
  })
  if (!res.ok) return null
  return (await res.json()) as VoiceSynthesizeResponse
}

export async function runVoicePipeline(body: VoicePipelineRequest): Promise<VoicePipelineResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/pipeline`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) return null
  return (await res.json()) as VoicePipelineResponse
}

export async function transcribeVoice(body: VoiceTranscribeRequest): Promise<VoiceTranscribeResponse | null> {
  const res = await fetch(`${BASE}/api/v1/voice/transcribe`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      audio_base64: body.audio_base64,
      audio_format: body.audio_format ?? 'webm',
      language: body.language ?? 'de-DE',
    }),
  })
  if (!res.ok) return null
  return (await res.json()) as VoiceTranscribeResponse
}
